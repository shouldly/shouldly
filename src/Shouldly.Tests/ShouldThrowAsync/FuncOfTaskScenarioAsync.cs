#if net40
using System;
using System.Threading;
using System.Threading.Tasks;
using Xunit;

namespace Shouldly.Tests.ShouldThrowAsync
{  
    public class FuncOfTaskScenarioAsync
   
    {

    [Fact]
        public void ShouldThrowAWobbly()
        {
            try
            {
                Task task = Task.Factory.StartNew(() => { var a = 1 + 1; Console.WriteLine(a); },
                        CancellationToken.None, TaskCreationOptions.None,
                        TaskScheduler.Default);

                var result = task.ShouldThrowAsync<InvalidOperationException>("Some additional context");
                result.Wait();
            }
            catch (AggregateException e)
            {
                var inner = e.Flatten().InnerException;
                var ex = inner.ShouldBeOfType<ShouldAssertException>();
                ex.Message.ShouldContainWithoutWhitespace(@"
                            `task` should throw System.InvalidOperationException but did not
                            Additional Info:
                            Some additional context");
            }
        }

        [Fact]
        public void ShouldPass()
        {
            var task = Task.Factory.StartNew(() => { throw new InvalidOperationException(); },
                CancellationToken.None, TaskCreationOptions.None,
                TaskScheduler.Default);

            var result = task.ShouldThrowAsync<InvalidOperationException>();
            result.Wait();
        }
    }
}
#endif