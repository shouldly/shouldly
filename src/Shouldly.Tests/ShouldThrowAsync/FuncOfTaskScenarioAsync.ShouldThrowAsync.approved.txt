Task `doSomething()`
    should throw
System.DivideByZeroException
    but did not